flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.executiongraph.failover.flip1;
+
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Result containing the tasks to restart upon a task failure.
+ * Also contains the reason if the failure is not recoverable(non-recoverable
+ * failure type or restarting suppressed by restart strategy).
+ */
+public class FailureHandlingResult {
+
+	/** Task vertices to restart to recover from the failure. */
+	private Set<ExecutionVertexID> verticesToRestart;
+
+	/** Delay before the restarting can be conducted. */
+	private long restartDelayMS;
+
+	/** Reason why the failure is not recoverable. */
+	private Throwable error;
+
+	/**
+	 * Creates a result of a set of tasks to restart to recover from the failure.
+	 *
+	 * @param verticesToRestart containing task vertices to restart to recover from the failure
+	 * @param restartDelayMS indicate a delay before conducting the restart
+	 */
+	public FailureHandlingResult(Set<ExecutionVertexID> verticesToRestart, long restartDelayMS) {
+		checkState(restartDelayMS >= 0);
+
+		this.verticesToRestart = checkNotNull(verticesToRestart);
+		this.restartDelayMS = restartDelayMS;
+	}
+
+	/**
+	 * Creates a result that the failure is not recoverable and no restarting should be conducted.
+	 *
+	 * @param error reason why the failure is not recoverable
+	 */
+	public FailureHandlingResult(Throwable error) {
+		this.error = checkNotNull(error);
+	}
+
+	/**
+	 * Returns the tasks to restart.
+	 *
+	 * @return the tasks to restart
+	 */
+	public Set<ExecutionVertexID> getVerticesToRestart() {
+		if (canRestart()) {
+			return verticesToRestart;
```

Comment:
return Collections.unmodifiableSet(verticesToRestart);

Reply:
I think it's not necessary to add extra cost to do this conversion.

Reply:
Instances of this class can be made immutable and I don't see a reason not to.

Reply:
Ok. Let's make it immutable.

Reply:
nit: we could make the `verticesToRestart` field immutable instead and return it without additional wrappers.

Reply:
Ok.